overlord/devicestate/users.go

```diff
@@ -0,0 +1,359 @@
+	st.Unlock()
+	// ErrInvalidUser means "not found" in this case
+	if err != nil && err != auth.ErrInvalidUser {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+	return u, nil
+}
+
+func CreateUser(st *state.State, mgr *DeviceManager, sudoer bool, createKnown bool, email string) ([]UserResponse, *UserError) {
+	var model *asserts.Model
+	var serial *asserts.Serial
+	if createKnown {
+		var err error
+		st.Lock()
+		model, err = mgr.Model()
+		st.Unlock()
+		if err != nil {
+			return nil, &UserError{Internal: true, Err: fmt.Errorf("cannot create user: cannot get model assertion: %v", err)}
+		}
+		st.Lock()
+		serial, err = mgr.Serial()
+		st.Unlock()
+		if err != nil && !errors.Is(err, state.ErrNoState) {
+			return nil, &UserError{Internal: true, Err: fmt.Errorf("cannot create user: cannot get serial: %v", err)}
+		}
+	}
+
+	st.Lock()
+	db := assertstate.DB(st)
+	st.Unlock()
+
+	u := &userManager{
+		state:    st,
+		safe:     false,
+		assertDb: db,
+		modelAs:  model,
+		serialAs: serial,
+		isSudoer: sudoer,
+	}
+
+	// special case: the user requested the creation of all known
+	// system-users
+	if email == "" && createKnown {
+		return u.createAllKnownSystemUsers()
+	}
+	if email == "" {
+		return nil, &UserError{Internal: true, Err: fmt.Errorf("cannot create user: 'email' field is empty")}
+	}
+
+	var username string
+	var opts *osutil.AddUserOptions
+	var err error
+	if createKnown {
+		username, opts, err = getUserDetailsFromAssertion(u.assertDb, u.modelAs, u.serialAs, email)
+	} else {
+		st.Lock()
+		storeService := snapstate.Store(st, nil)
+		st.Unlock()
+		username, opts, err = getUserDetailsFromStore(storeService, email)
+	}
+
+	if err != nil {
+		return nil, &UserError{Internal: false, Err: fmt.Errorf("%s", err)}
+	}
+
+	opts.Sudoer = sudoer
+	opts.ExtraUsers = !release.OnClassic
+	createdUser, err := u.addUser(username, email, opts)
+	if err != nil {
+		return nil, &UserError{Internal: true, Err: err}
+	}
+
+	var createdUsers []UserResponse
+	createdUsers = append(createdUsers, createdUser)
+	return createdUsers, nil
+
+}
+
+func getUserDetailsFromStore(theStore snapstate.StoreService, email string) (string, *osutil.AddUserOptions, error) {
+	v, err := theStore.UserInfo(email)
+	if err != nil {
+		return "", nil, fmt.Errorf("cannot create user %q: %s", email, err)
+	}
+	if len(v.SSHKeys) == 0 {
+		return "", nil, fmt.Errorf("cannot create user for %q: no ssh keys found", email)
+	}
+
+	// Amend information where the key came from to ensure it can
+	// be update/replaced later
+	for i, k := range v.SSHKeys {
+		v.SSHKeys[i] = fmt.Sprintf(`%s # snapd {"origin":"store","email":%q}`, k, email)
+	}
+
+	gecos := fmt.Sprintf("%s,%s", email, v.OpenIDIdentifier)
+	opts := &osutil.AddUserOptions{
+		SSHKeys: v.SSHKeys,
+		Gecos:   gecos,
+	}
+	return v.Username, opts, nil
+}
+
+// userManager is helper to handle safe and unsafe modes of operations
+type userManager struct {
+	state    *state.State
+	safe     bool
```

Comment:
This needs to be explained better. Given that the default value for booleans is `false`, I'd actually suggest renaming it to `unsafe` and switch the conditions.

I'm especially confused because I see that we do the lock of the state when `unsafe` is `false`, which at first sight seems counter-intuitive: the safe way of operating is generally with locking...

Reply:
I'm also a bit confused by this. The `safe` field is also only ever set to `false` so it seems like it's not needed and the functions that used it could be simplified. 
Another thing (more related to structure) is that having this as a `userManager` seems odd since we create one at every `CreateUser` call, it's not long-lived and the values of the fields are never overwritten after creation. It acts more like an options struct carrying data. Maybe we could follow path and have something like: 
```
type createUserOpts struct {
	assertDb asserts.RODatabase
	modelAs  *asserts.Model
	serialAs *asserts.Serial
	isSudoer bool
}

func createAllKnownSystemUsers(st *state.State, opts *createUserOpts) ([]UserResponse, *UserError) {
    ...
}
```
I removed the `safe` flag and moved the `state` variable out of it because it's common in for our exported functions `overlord` to take the state as a first argument.

Reply:
I remove `safe` variable from the struct, as we only run in unsafe more, so indeed there was no use for this switch.
I also remove state as proposed and moved it to the function argument